packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeActiveState.tsx

```diff
@@ -0,0 +1,100 @@
+import { useStableProps } from './useStableProps';
+import { BaseFlatTreeItem } from './flattenTree';
+import * as _ from 'lodash';
+
+export interface UseTreeActiveStateResult {
+  activeItemIds: string[];
+  toggleItemActive: (e: React.SyntheticEvent, idToToggle: string) => void;
+  expandSiblings: (e: React.KeyboardEvent, focusedItemId: string) => void;
+}
+
+export function useTreeActiveState(
+  props: Pick<UseTreeOptions, 'defaultActiveItemIds' | 'activeItemIds' | 'exclusive'>,
+  getItemById: (id: string) => BaseFlatTreeItem,
+  deprecated_initialActiveItemIds: string[],
+): UseTreeActiveStateResult {
+  const [activeItemIds, setActiveItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultActiveItemIds,
+    value: props.activeItemIds,
+    initialValue: deprecated_initialActiveItemIds,
+  });
+
+  const stableProps = useStableProps(props);
+
+  const toggleItemActive = React.useCallback(
+    (e: React.SyntheticEvent, idToToggle: string) => {
+      const item = getItemById(idToToggle);
+      if (!item || !item.hasSubtree) {
+        // leaf node does not have the concept of active/inactive
+        return;
+      }
+
+      setActiveItemIdsState(activeItemIds => {
+        let nextActiveItemIds;
+        const index = activeItemIds.indexOf(idToToggle);
+
+        if (index >= 0) {
+          nextActiveItemIds = _.without(activeItemIds, idToToggle);
+        } else if (props.exclusive) {
+          // need to collapse everything else, except id and its ancestors
+          const ancestors = getAncestorsIds(getItemById, item);
+          nextActiveItemIds = [...ancestors, idToToggle];
+        } else {
+          nextActiveItemIds = [...activeItemIds, idToToggle];
+        }
+
+        _.invoke(stableProps.current, 'onActiveItemIdsChange', e, {
+          ...stableProps.current,
+          activeItemIds: nextActiveItemIds,
+        });
+
+        return nextActiveItemIds;
+      });
+    },
+    [getItemById, props.exclusive, setActiveItemIdsState, stableProps],
+  );
+
+  const expandSiblings = React.useCallback(
+    (e: React.KeyboardEvent, focusedItemId: string) => {
+      const item = getItemById(focusedItemId);
+      if (!item) {
+        return;
+      }
+
+      const siblingsIds = _.without(getItemById(item?.parent)?.childrenIds || [], focusedItemId);
+
+      if (!siblingsIds) {
+        return;
+      }
+
+      setActiveItemIdsState(activeItemIds => {
+        const nextActiveItemIds = _.uniq(activeItemIds.concat(siblingsIds));
+        _.invoke(stableProps.current, 'onActiveItemIdsChange', e, {
+          ...stableProps.current,
+          activeItemIds: nextActiveItemIds,
+        });
+        return nextActiveItemIds;
+      });
+    },
+    [getItemById, setActiveItemIdsState, stableProps],
+  );
+
+  return {
+    activeItemIds,
+    toggleItemActive,
+    expandSiblings,
+  };
+}
+
+function getAncestorsIds(getItemById: (id: string) => BaseFlatTreeItem, item: BaseFlatTreeItem): string[] {
```

Comment:
Question: 
`GetItemById` would be in defined `useTree.tsx` (because `useGetItemById` was moved into `useTree`)
So there's circular reference:
`useTree` imports `useTreeActiveState`
`useTreeActiveState` imports `GetItemById` exported by `useTree`.
How do I deal with it? Should I create a separate type file?